### PR TITLE
feat: add support for on-cluster build (#200)

### LINKED ISSUE #200: Investigate on how to provide support for build on cluster feature
Building a function on cluster is an operation which requires tekton and knative func cli.

The process is quite simple (starting from an empty cluster)
1) install tekton on your cluster
2) install all tekton tasks needed for building a knative func + add permissions to the default service account
3) create a new func
4) initialize a git repo 
5) add the build git section to the `func.yaml` file. Example:
```
build: git                                          # required, specify `git` build type
git:
  url: https://github.com/my-repo/my-function.git   # required, git repository with the function source code
  revision: main                                    # optional, git revision to be used (branch, tag, commit)
  # contextDir: myfunction                          # optional, needed only if the function is not located
                                                    # in the repository root folder
```
6) push everything on git
7) execute the `kn func deploy` command
8) to update your func, push the new code to git and re-run `kn func deploy`

So what kind of support we would like to offer? 
By separating each point I think this. The only aspect I can think of (not sure if and how it can be possible) is the interaction with the tekton plugin if installed.

1) nothing to do
2) nothing? The tekton tasks installation is something that can be done with the tekton plugin. Nothing related to knative.
3) already supported
4) nothing to do
5) we can add a code completion helper when working with it
6) nothing to do
7) already supported
8) nothing to do

Thoughts? WDYT? @mohitsuman @sudhirverma 
Description -> https://github.com/knative-sandbox/kn-plugin-func/blob/044177026725adb6591c7cf4593d59755e34dd0a/docs/reference/on_cluster_build.md

### PR DESCRIPTION
**Pull request by lstocchi**
it resolves #200 
it needs https://github.com/redhat-developer/intellij-common/pull/165 to work

This extend the deploy action to be run with the `--remote` and `--git-url` flag and support the on-cluster build action.

The feature itself is still a bit unstable from my pov. It failed easily if the env is not set correctly.

What i do to setup it is: start a cluster with clusterBot, install the pipeline 1.8 and serverless operators, create a new project,  install the following tasks
```
REPO=https://raw.githubusercontent.com/openshift-knative/kn-plugin-func/serverless-1.25.0
oc apply -f ${REPO}/pipelines/resources/tekton/task/func-buildpacks/0.1/func-buildpacks.yaml
oc apply -f ${REPO}/pipelines/resources/tekton/task/func-s2i/0.1/func-s2i.yaml
oc apply -f https://gist.githubusercontent.com/jrangelramos/15cad16980957c19c621e37ec8094cb4/raw/a4007c4defffc35d1da743b5994e7eb372506a05/func-deploy.yaml

```

and the git-clone task 1.7 from the tektonhub
```
kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/main/task/git-clone/0.7/git-clone.yaml
```

then everything works fine.

I've called the action `on-cluster build` as found in the serverless documentation but it performs a build and deploy action so not sure if we want to rename it for clarity.

on-cluster-build

### CODE REVIEW

File: src/main/java/com/redhat/devtools/intellij/knative/actions/func/BuildAction.java [hunk 1]
```diff
@@ -191,22 +175,27 @@ protected void doExecuteAction(Project project, Function function, String regist
...
 
-    protected String getImageFromUser(String name) {
+    protected ImageRegistryModel getImageFromUser(String name) {
         String defaultUsername = System.getProperty("user.name");
         String defaultImage = "quay.io/" + defaultUsername + "/" + name + ":latest";
-        Messages.InputDialog dialog = new Messages.InputDialog(null, "Provide full image name in the form [registry]/[namespace]/[name]:[tag] (e.g quay.io/boson/image:latest)",
-                "Build Function " + name, null, defaultImage,
+        InputDialogWithCheckbox dialog = new InputDialogWithCheckbox("Provide full image name in the form [registry]/[namespace]/[name]:[tag] (e.g quay.io/boson/image:latest)",
+                "Build Function " + name,
+                "Autodiscovery an internal registry and use it to deploy",
```
Comment (mohitsuman) [inline src/main/java/com/redhat/devtools/intellij/knative/actions/func/BuildAction.java:194]:
```suggestion
                "Auto discover an internal registry and use it to deploy",
```

**Review by mohitsuman — CHANGES_REQUESTED** [review]
LGTM, just one small change with space in the word.

File: src/main/java/com/redhat/devtools/intellij/knative/actions/func/BuildAction.java [hunk 2]
```diff
@@ -191,22 +175,27 @@ protected void doExecuteAction(Project project, Function function, String regist
...
         return true;
     }
 
-    protected String getImageFromUser(String name) {
+    protected ImageRegistryModel getImageFromUser(String name) {
         String defaultUsername = System.getProperty("user.name");
         String defaultImage = "quay.io/" + defaultUsername + "/" + name + ":latest";
-        Messages.InputDialog dialog = new Messages.InputDialog(null, "Provide full image name in the form [registry]/[namespace]/[name]:[tag] (e.g quay.io/boson/image:latest)",
-                "Build Function " + name, null, defaultImage,
+        InputDialogWithCheckbox dialog = new InputDialogWithCheckbox("Provide full image name in the form [registry]/[namespace]/[name]:[tag] (e.g quay.io/boson/image:latest)",
```
Comment (jeffmaury) [inline src/main/java/com/redhat/devtools/intellij/knative/actions/func/BuildAction.java:192]: [] are generally used to denote optional parts so maybe should be `registry/namespace/name[:tag]`

**Review by mohitsuman — APPROVED** [review]

**Comment by lstocchi** [comment]
> I think this needs to be better integrated with the Git plugin as:
> 
> * we should be able to auto detect the Git repo URL
> * In case the project is not configured on Git we should ask the user to ensure
> * If the project is configured on Git and we detect local changes then we should tell it to the user and if possible (Git plugin integration) push the changes
> * Don't understand why if I select auto discovery of registry I can't change the image name ?
> * I think we should at least check that Tekton is installed on the cluster

@jeffmaury Just talked with Mohit about the release of Ij and Vscode. As we are in a hurry (this week serverless func is going GA and we should release our extensions as well) he proposed to create a new issue with all your comments and work on it in the next sprint. So, talking about what we have now (a basic workflow where the user needs to add his gh repo), is it ok or you have any concern? Thanks!!

**Comment by jeffmaury** [comment]
No why are we in a hurry ? I would strongly object against this PT being merged because I think it adds very little value to the user. We are not required to have the plugin being released in full sync with the product.

**Comment by lstocchi** [comment]
> No why are we in a hurry ? I would strongly object against this PT being merged because I think it adds very little value to the user. We are not required to have the plugin being released in full sync with the product.

Even though i agree with you, in the sense that i find this feature not so useful from an IDE perspective, i don't see how leveraging the git plugin would add any more value to the workflow from what it has now. Personally, i don't use the git plugin in IJ. I always perform my stuff on the terminal. So it would be a pain for me to being forced to use it.

Regarding being in a hurry, @mohitsuman could certainly explain it better :laughing:

**Comment by jeffmaury** [comment]
> > No why are we in a hurry ? I would strongly object against this PT being merged because I think it adds very little value to the user. We are not required to have the plugin being released in full sync with the product.
> 
> Even though i agree with you, in the sense that i find this feature not so useful from an IDE perspective, i don't see how leveraging the git plugin would add any more value to the workflow from what it has now. Personally, i don't use the git plugin in IJ. I always perform my stuff on the terminal. So it would be a pain for me to being forced to use it.
> 
> Regarding being in a hurry, @mohitsuman could certainly explain it better 😆

I don't use it but I think if your project has git metadata then we should retrieved it and monitor the state (so probably using the Git plugin will help)

**Comment by mohitsuman** [comment]
There is no hurry for the release. Anyway, the serverless GA is postponed as some testing is needed. From the IDE side, let's try to make sure we have the plugin stable and then release it. We can investigate if Git plugin integration helps the use case or not.

**Comment by lstocchi** [comment]
@jeffmaury I think i covered everything except helping users to push uncommitted changes.
I just added a series of warning messages based on the situation because then it really depends on what they want to do. Maybe they want to build another branch for some reason. 

A valid branch is selected
image

A branch that exists locally but not remotely
image

When there are uncommitted changes
image

A project which is not a git repo
image

**Review by jeffmaury — APPROVED** [review]
LGTM

File: src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        addBranchField(currentRemoteAndBranch.getSecond());
+
+        JPanel panel = new JPanel(new BorderLayout());
+        contentPanel.add(panel);
+    }
+
+    private void addTopWarningMessage(String remote) {
+        String message = "";
+        if (gitRepository == null) {
+            message = "Your project is not a git repository. Please initialize it before to proceed building it on cluster.";
```
Comment (mohitsuman) [inline src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java:113]:
```suggestion
            message = "This project is not a git repository. Please git initialise it and then proceed to build on the cluster.";
```

File: src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        } else {
+            List<GitFileStatus> gitFileStatuses = gitRepository.getStagingAreaHolder().getAllRecords();
+            if (remote.isEmpty()) {
+                message = "Your local branch is not present remotely. Push it before to proceed building it on cluster";
```
Comment (mohitsuman) [inline src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java:117]:
```suggestion
                message = "The local branch is not present remotely. Push it to remote and then proceed to build it on cluster";
```

File: src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            } else if (!gitFileStatuses.isEmpty()) {
+                message = "Your local branch contains some uncommitted changes. Push them before to proceed building it on cluster";
```
Comment (mohitsuman) [inline src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java:119]:
```suggestion
                message = "The local branch contains some uncommitted changes. Push those changes and then proceed to build it on cluster";
```

File: src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+    private void addTopDescription(String descriptionText) {
+        JLabel topDescription = new JLabel(descriptionText);
+        topDescription.setBorder(JBUI.Borders.empty(10, 0, 10, 10));
+        addComponentToContent(contentPanel, null, topDescription, null, 5);
+    }
+
+    private void addRepoField(String remote) {
+        JLabel lblIDParam = createLabel("Repo:",
+                "The repository to use for building and deploying.",
```
Comment (mohitsuman) [inline src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java:155]:
```suggestion
                "Provide git repository with the function source code",
```

File: src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        panelGitRepoError.add(lblGitRepoError);
+        panelGitRepoError.setAlignmentX(Component.LEFT_ALIGNMENT);
+        panelGitRepoError.setVisible(false);
+
+        contentPanel.add(panelGitRepoError);
+    }
+
+    private void addBranchField(String branch) {
+        JLabel lblBranchParam = createLabel("Branch:",
+                "The branch to use for building and deploying. If left empty, the main branch is used.",
```
Comment (mohitsuman) [inline src/main/java/com/redhat/devtools/intellij/knative/ui/GitDialog.java:193]:
```suggestion
                "Git revision to be used (branch, tag, commit). If left empty, the main branch is used.",
```

File: src/main/resources/func.json [hunk 8]
```diff
@@ -1,30 +1,30 @@
 {
   "tools": {
     "func": {
-      "version": "0.26.0",
+      "version": "1.7.0",
```
Comment (mohitsuman) [inline src/main/resources/func.json:4]: @lstocchi For func cli, 1.7.1 is also out. We can migrate to that version.

**Review by mohitsuman — CHANGES_REQUESTED** [review]
Few wording changes.

File: src/main/resources/func.json (hunk 8 above)
Comment (lstocchi) [reply src/main/resources/func.json:4]: I'm going to create a new issue to bump it.

**Comment by lstocchi** [comment]
Covered all Mohit's suggestions + add a check to see if the git plugin is enabled

**Review by mohitsuman — APPROVED** [review]
LGTM

**Pull request merged**